### PR TITLE
Packet integrity

### PR DESCRIPTION
**Pull request by lkrawczyk**

### CODE REVIEW

File: dev/src/base/net/src/codec.cpp [hunk 1]
```diff
@@ -55,17 +57,18 @@ namespace base
...
-                write(&o, sizeof(o));
+                return static_cast<T *>(write(&o, sizeof(o)));
             }
 
-            void write(const void* data, Uint32 size)
+            void* write(const void* data, Uint32 size)
             {
                 FATAL_ASSERT(size + m_size <= m_capacity);
                 Uint8* destination = m_buffer + m_size;
                 mem::ACopyMem(destination, data, size);
```
Comment (rexdex) [inline dev/src/base/net/src/codec.cpp]: memcpy will be ultra slow with small data sizes.

File: dev/src/base/net/src/codec.cpp [hunk 2]
```diff
@@ -108,13 +120,19 @@ namespace base
                 ackHeader.m_fragmentIndex = packet->getFragmentIndex();
 
                 writer.write(&ackHeader, sizeof(ackHeader));
+
+                crc.append(&ackHeader, sizeof(ackHeader));
             }
 
+            // Patch header that was written at the beginning with CRC checksum
+            // truncated to 24 bits
+            writtenHeader->setChecksum(crc.getCRC() % (1 << 24));
```
Comment (rexdex) [inline dev/src/base/net/src/codec.cpp]: hm, a little bit of abstraction leakage - we adjust what we store based on data that we should not know (that the field is 24 bit)

File: dev/src/base/net/src/codec.cpp [hunk 3]
```diff
@@ -141,7 +161,12 @@ namespace base
...
-                packet->writeData(reader.read(dataHeader->m_dataSize), dataHeader->m_dataSize);
+                auto data = reader.read(dataHeader->m_dataSize);
+
+                packet->writeData(data, dataHeader->m_dataSize);
+
+                crc.append(dataHeader, sizeof(DataPacketHeader));
+                crc.append(data, dataHeader->m_dataSize);
             }
 
             if (packet->getType() == PacketType::DataAcknowledge)
```
Comment (rexdex) [inline dev/src/base/net/src/codec.cpp]: else if

File: dev/src/base/net/src/codec.cpp [hunk 4]
```diff
@@ -150,10 +175,18 @@ namespace base
 
                 packet->setSequenceNumber(dataAckHeader->m_sequenceNumber);
                 packet->setFragmentIndex(dataAckHeader->m_fragmentIndex);
+
+                crc.append(dataAckHeader, sizeof(DataAcknowledgeHeader));
             }
```
Comment (rexdex) [inline dev/src/base/net/src/codec.cpp]: else - invalid packet type

File: dev/src/base/net/src/codec.cpp [hunk 5]
```diff
@@ -150,10 +175,18 @@ namespace base
...
 
-            FATAL_ASSERT(reader.getRemaining() == 0);
+            // Calculate checksum for received data
+            Uint32 checksum = crc.getCRC() % (1 << 24);
```
Comment (rexdex) [inline dev/src/base/net/src/codec.cpp]: similar abstraction leakage as at the place it was stored

File: dev/src/base/net/src/connection.cpp [hunk 6]
```diff
@@ -22,11 +22,8 @@ namespace base
         {
             auto index = m_idAllocator.alloc();
             if (index == INVALID_ID)
-            {
                 return INVALID_CONNECTION_ID;
```
Comment (rexdex) [inline dev/src/base/net/src/connection.cpp]: I would probably log this, something like TRACE_ERROR("Net: Unable to allocate connection ID"); etc

File: dev/src/base/net/src/socket.cpp [hunk 7]
```diff
@@ -159,7 +159,7 @@ namespace base
                 // Writing to socket failed, exit with error         
                 if (bytesSent < 0)
                     return false;
-            } while (bytesSent != static_cast<Int32>(bufferSize));
+            } while (bytesSent != range_cast<Int32>(bufferSize));
```
Comment (rexdex) [inline dev/src/base/net/src/socket.cpp]: this is potentially costly, range_cast is intended for places where values is assigned

File: dev/src/base/net/src/codec.cpp (hunk 4 above)
Comment (lkrawczyk) [reply dev/src/base/net/src/codec.cpp]: This packet type list is not exhaustive - there are other packet types that don't need additional decoding. In the future most likely packets of all types will carry some additional data, I will add this terminal check then.

**Pull request merged**